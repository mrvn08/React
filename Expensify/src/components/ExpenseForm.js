import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

const now = moment();
console.log(now.format('MMM Do, YYYY'));

export default class ExpenseForm extends React.Component {
    constructor(props){
        super();

        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount/100).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: ''
        };
    }

    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description }));
    };

    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({ note }));
    };

    onDateChange = (createdAt) => {
        if(createdAt){
            this.setState(() => ({ createdAt }));
        }
    };

    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }));
    };

    onAmountChange = (e) => {
        const amount = e.target.value;

        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){
            this.setState(()=>({ amount }));
        }
    }

    onSubmit = (e) => {
        e.preventDefault();
        if(!this.state.description || !this.state.amount){
            this.setState(() => ({ error: 'Please enter a value in Description and amount' }));
        } else {
            this.setState(() => ({ error: '' }));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                note: this.state.note,
                createdAt: this.state.createdAt.valueOf()
            })
        }
    };

    render() {
        return (
            <div>
                <form onSubmit={this.onSubmit}>
                    <input 
                        type="text" 
                        placeholder="Description"
                        autoFocus
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                    />
                    <input
                        type="text"
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />
                    <SingleDatePicker
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                    />
                    <textarea
                        type="note"
                        placeholder="Add a note for this expense (optional)"
                        onChange={this.onNoteChange}
                        value={this.state.note}
                    />
                    {this.state.error && <p>{this.state.error}</p>}
                    <button>Add Expense</button>
                </form>
            </div>
        );
    }
}

