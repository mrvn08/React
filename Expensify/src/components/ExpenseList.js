import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

export const ExpenseList = (props) => (
    <div>
        <h1>Expense List</h1>
        {
            props.expenses.length === 0 ? (
                <p>No expenses</p>
            ) : props.expenses.map((expense, index) => <ExpenseListItem key={expense.id} {...expense}/>)            
        }
    </div>
);

//Unconnected Expense List is exported as a named export for testing

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    };
};

//Automatically reruns whenever state in store is changed
//state.expenses is being passed to prop expenses in ExpenseList
//When component is connected to store, it is reactive. The values change on screen realtime

export default connect(mapStateToProps)(ExpenseList);

