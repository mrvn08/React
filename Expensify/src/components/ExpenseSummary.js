import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import getVisibleExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';

export const ExpenseSummary = (props) => (
    <div>
        <p>{props.numExpenses > 0 ? `Viewing ${props.numExpenses} ${props.numExpenses > 1 ? "expenses" : "expense"} totalling ${numeral(props.totalExpenses).format('$0,0.00')}` : `No expenses to view`}</p>
    </div>
);

const mapStateToProps = (state) => ({
    numExpenses: getVisibleExpenses(state.expenses, state.filters).length,
    totalExpenses: selectExpensesTotal(getVisibleExpenses(state.expenses, state.filters))/100,
});

export default connect(mapStateToProps)(ExpenseSummary);
