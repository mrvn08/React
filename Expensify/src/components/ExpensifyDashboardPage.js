import React from 'react';

const ExpenseDashboardPage = () => (
    <div>
        This is a dashboard component
    </div>
);

export default ExpenseDashboardPage;
