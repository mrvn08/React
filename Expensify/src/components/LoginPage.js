import React from 'react';
import { connect } from 'react-redux';
import { startLogin } from '../actions/auth';

export const LoginPage = ({ startLogin }) => (
    <div>
        <form>
            <h4>Expensify</h4>
            <input type="text" name="username" placeholder="Username"/>
            <input type="password" name="password" placeholder="Password"/>
            
        </form>
        <button onClick={startLogin} >Login</button>
    </div>
);

const mapDispatchToProps = (dispatch) => ({
    startLogin: () => dispatch(startLogin())
});

export default connect(undefined, mapDispatchToProps)(LoginPage);