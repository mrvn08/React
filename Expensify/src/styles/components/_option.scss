.option{
    border-bottom: 1px solid lighten($light-blue, 10%);
    display: flex;
    justify-content: space-between;
    padding: $medium $regular;
}

.option__text{
    color: white;
    font-weight: 500;
    font-size: 2rem;
    margin: 0;
    margin-right: $small;
    word-break: break-all;
}
