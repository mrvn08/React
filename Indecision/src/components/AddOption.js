//remake with babel transform class properties
import React from 'react';

export default class AddOption extends React.Component{
    state = {
        error: undefined
    };

    handleAddOption = (e) => {
        e.preventDefault();
        const option = e.target.elements.option.value.trim();
        const error = this.props.handleAddOption(option);

        this.setState(() => ({ error }));

        if(!error){
            e.target.elements.option.value = "";
        }
    }

    //We're still going to use regular es6 for react methods such as render() and react component lifecycle methods.
    render(){
        return (
            <div>
                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.handleAddOption}>
                    <input name="option" type="text" placeholder="C'mooooon do it! DO IT NOW!"/>
                    <button>Add Option</button>
                </form>
            </div>
        );
    }
}