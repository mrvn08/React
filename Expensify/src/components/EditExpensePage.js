import React from 'react';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';
import { connect } from 'react-redux';

export class EditExpensePage extends React.Component{
    onSubmit = (expense) => {
        this.props.startEditExpense(this.props.match.params.id, expense);
        this.props.history.push('/');
    };

    onClick = () => {
        this.props.startRemoveExpense({ id: this.props.match.params.id });
        this.props.history.push('/');
    };

    render() {
        return (
            <div>
                <ExpenseForm 
                    expense={this.props.expense}
                    onSubmit={this.onSubmit}
                />
                <button onClick={this.onClick}>Remove</button>
            </div>
        );
    }
}

const mapStateToProps = (state, props) => ({
    expense: state.expenses.find((expense) => expense.id == props.match.params.id )
});

const mapDispatchToProps = (dispatch) => ({ 
    startEditExpense: ( id, expense ) => dispatch(startEditExpense(id, expense)),
    startRemoveExpense: ({ id }) => dispatch(startRemoveExpense({ id: id }))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
