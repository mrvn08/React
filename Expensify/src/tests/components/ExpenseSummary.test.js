import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseSummary } from '../../components/ExpenseSummary';

test('Should render no expense to view message', () => {
    const wrapper = shallow(<ExpenseSummary numExpenses = {0} />);
    expect(wrapper).toMatchSnapshot();
});

test('Should render expenses and their total', () => {
    const wrapper = shallow(<ExpenseSummary numExpenses = {2} totalExpenses = {123.12} />);
    expect(wrapper).toMatchSnapshot();
});

test('Should render an expense and its total', () => {
    const wrapper = shallow(<ExpenseSummary numExpenses = {1} totalExpenses = {50.4} />);
    expect(wrapper).toMatchSnapshot();
});
