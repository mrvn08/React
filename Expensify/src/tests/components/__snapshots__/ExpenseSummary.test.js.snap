// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render an expense and its total 1`] = `
<div>
  <p>
    Viewing 1 expense totalling $50.40
  </p>
</div>
`;

exports[`Should render expenses and their total 1`] = `
<div>
  <p>
    Viewing 2 expenses totalling $123.12
  </p>
</div>
`;

exports[`Should render no expense to view message 1`] = `
<div>
  <p>
    No expenses to view
  </p>
</div>
`;
